### ISSUE TITLE
TranslationLoader deprecation message in Symfony 3.4.2

### ISSUE THREAD

**Issue by iquito** [symfony/symfony#25518]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.4.2

In my application I get the following deprecation message:

    The class "Symfony\Bundle\FrameworkBundle\Translation\TranslationLoader" is deprecated since version 3.4 and will be removed in 4.0. Use "Symfony\Component\Translation\Reader\TranslationReader" instead. 

I am using translations, and also JMSTranslationBundle as a helper (to extract annotations and generate .xlf files). But TranslationLoader or anything like it is not used anyhere in my code base. I am basically using a very standard setup for translations, with `translator:      { fallbacks: [de] }` and .xlf files. It is the most basic setup I can imagine.

This is the output in Symfony Profiler for the error:

    [▼
        "exception" => SilencedErrorContext {#1537 ▼
            +count: 1
            -severity: E_USER_DEPRECATED
            trace: {▼
            /projectdir/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Translation/TranslationLoader.php:17 {▶}
            /projectdir/vendor/symfony/symfony/src/Symfony/Component/Debug/DebugClassLoader.php:143 {▶}
            }
        }
    ]

    {▼
        /projectdir/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Translation/TranslationLoader.php:17 {▼
            › 
            › @trigger_error(sprintf('The class "%s" is deprecated since version 3.4 and will be removed in 4.0. Use "%s" instead. ', TranslationLoader::class, TranslationReader::class), E_USER_DEPRECATED);
            › 
        }
        /projectdir/vendor/symfony/symfony/src/Symfony/Component/Debug/DebugClassLoader.php:143 {▼
            › if ($file = $this->classLoader[0]->findFile($class)) {
            ›     require $file;
            › }
            arguments: {▼
            "/projectdir/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Translation/TranslationLoader.php"
            }
        }
    }

I do not know where the actual usage of TranslationLoader originates, or how to find out, because the stack trace only goes back to DebugClassLoader. So I am not sure if this is a bug or just some other problem. Maybe someone knows how to actually get to the root of this problem, or how to find out more.

**Comment by nicolas-grekas** [comment]
Does the issue go away when you remove `JMSTranslationBundle`? If yes, I'd suggest to open an issue there instead. Otherwise, can you please provide [a short reproducer](https://symfony.com/doc/current/contributing/code/reproducer.html)?

**Comment by xabbuh** [comment]
I am closing here as this issue does seem to be related to JMSTranslationBundle. But we can of course reopen here if there is evidence that this behaviour is related to Symfony itself.

**Comment by iquito** [comment]
I tried it out without JMSTranslationBundle - the problem still persists. It occurs on any page of my project, no matter if any translations are loaded/used or not.

Basically, somewhere somehow both `Symfony\Bundle\FrameworkBundle\Translation\TranslationLoader` and `Symfony\Component\HttpFoundation\Session\Storage\Handler\WriteCheckSessionHandler` classes are loaded (WriteCheckSessionHandler is seemingly loaded also because of the inclusion of TranslationLoader), triggering the deprecation notices, although both classes are not used by me, and I do not know where it would be used, or how.

At the same time, it is very difficult to find that out: The deprecation messages only go back to `DebugClassLoader`, where the class is loaded, but not further back to show where the class was used. So right now I would need to guess where it is used, or in what way, which could take forever. I invested about 4 hours recently to find the cause without any success, also trying to somehow get a full stack trace for the deprecation warnings, but I did not succeed.

I am only using the following bundles:

    Symfony\Bundle\FrameworkBundle\FrameworkBundle()
    Symfony\Cmf\Bundle\RoutingBundle\CmfRoutingBundle()
    Symfony\Bundle\TwigBundle\TwigBundle()
    Symfony\Bundle\MonologBundle\MonologBundle()
    Doctrine\Bundle\DoctrineBundle\DoctrineBundle()
    Sensio\Bundle\FrameworkExtraBundle\SensioFrameworkExtraBundle()
    Knp\Bundle\MenuBundle\KnpMenuBundle()
    Aws\Symfony\AwsBundle()
    Symfony\Bundle\DebugBundle\DebugBundle()
    Symfony\Bundle\WebProfilerBundle\WebProfilerBundle()

That is also why I am confused - my Symfony usage is quite vanilla, and my translation usage is even more basic and boring. Nothing to indicate why anything would trigger these deprecation notices. All bundles are at their highest possible version for Symfony 3.4.

**Comment by xabbuh** [comment]
Can you create a small example project that allows us to reproduce your issue?

**Comment by iquito** [comment]
I can try. That is probably a lengthy process though, which I cannot do right now. Is there any way of getting a full stack trace of a deprecation in Symfony? That would be so much easier.

**Comment by urg** [comment]
I'm having a similar issue - just tracked it down to this line: https://github.com/symfony/framework-bundle/blob/3.4/DependencyInjection/FrameworkExtension.php#L1121 - commenting it out removed the warning for me.

**Comment by iquito** [comment]
That doesn't do it for me - still occurs. Would also be weird if it was solved by that, because that does not actually use the service, just sets it to private for the service container. Somewhere the TranslationLoader has to be used & included for the error to occur.

**Comment by stedekay** [comment]
Also getting the same deprecation message and like @urg it will go away after commenting out line 1121 of the FrameworkExtension.

@iquito: Did you cleared your cache after commenting out that line? Because without clearing the cache the message still appeared for me.

**Comment by chalasr** [comment]
Unable to reproduce. Please provide the minimal steps to reproduce (starting from project creation).

**Comment by xabbuh** [comment]
I am closing for now. We can consider to reopen when we have a reproducable example.

**Comment by tealborder** [comment]
Im away from my computer but i had this issue when installing php5.6 onto debian jessie
The workaround for me was to manually specify a version number etc. Im just trying to keep this ticket alive. Will check back later with reproducible steps

**Comment by iquito** [comment]
I will attempt to upgrade to Symfony 4 and see if I get a clearer error there that I can report back - as soon as `symfony-cmf/routing-bundle` supports Symfony 4, which is keeping me from upgrading so far.

**Comment by richard-ejem** [comment]
For example, JMS AOP Bundle causes this too. The deprecation notice is emitted every time the TranslationLoader file is loaded by PHP, therefore it will happen every time when some script in the application tries to browse all services and eventually load their classes. 

The JMS AOP bundle's troublesome part is there: https://github.com/schmittjoh/JMSAopBundle/blob/master/DependencyInjection/Compiler/PointcutMatchingPass.php#L116 
It just checks for existence of each service's class, causing the `TranslationLoader.php` file to load and emit deprecation warning.

I don't like emitting a deprecation this way, maybe doing it in a constructor or so would be more friendly.

A simple solution I used to get rid of the warning is to remove the `translation.loader` service from the container - this can be done in your `Bundle::build`:

```php
class MyAppBundle extends Bundle
{
    public function build(ContainerBuilder $container)
    {
        $container->addCompilerPass(new class implements CompilerPassInterface {
            public function process(ContainerBuilder $container)
            {
                $container->removeDefinition('translation.loader');
            }
        }, PassConfig::TYPE_REMOVE);
    }
}
```

P.S. to see full stack trace of the warning, you can temporarily place a simple snippet just above the `trigger_error` call:

```php
debug_print_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS);die;
```

I find it annoying too that the profiler limits backtrace to 2 entries in this case, making it actually useless.

**Comment by iquito** [comment]
Thanks @richard-ejem! That enabled me to easily find the culprit - it was a compiler pass which did the following:

    foreach ($containerBuilder->getDefinitions() as $definition) {
      if (is_subclass_of($definition->getClass(), AbstractRouteCollectionProvider::class)) {
        $definition->addMethodCall('setLoaderResolver', [new Reference('routing.resolver')]);
      }
    }

I originally copied this from a library, so classes originating from a certain abstract class are automatically loaded as route resolvers. By calling `is_subclass_of` the class is loaded, even if it is never used after that (and TranslationLoader is never actually loaded for real page requests). I suspect something like this could also affect other libraries & code which go through the service container definitions. So these deprecation warnings are basically false-positives which would automatically disappear when switching to Symfony 4.

So in short, the current way of adding trigger_error just when the class is loaded has the chance of issuing false-positives, even if the class is never used and there is no deprecation problem. As @richard-ejem  mentioned, adding the deprecation warning in the class constructor might avoid such problems and make the deprecation warnings more meaningful.

**Comment by xabbuh** [comment]
I may oversee some discussion we had in the past about this, but this sounds reasonable to me.
